### PR TITLE
Per tile stats

### PR DESCRIPTION
**Pull request by rahwang**
For #5156. Not ready yet! Is this on the right track, @lilleyse ? Seems wrong to just be using the tileset._statistics for draw command, texture memory and vertex memory numbers, but I'm not sure how to compute these per tile.

### CODE REVIEW

**Comment by lilleyse** [comment]
Yes this is on the right track. I didn't look too deeply at the code yet, but just a few things:
* You can access the memory stats from the tile's content
* Now that triangles length is reported, also add that to tile info (add points length as well)
* Use the new depth-test off feature so labels aren't clipped https://github.com/AnalyticalGraphicsInc/cesium/pull/5166
* Use a smaller font, or adjust based on the number of properties showing
* Maybe this is already on your radar, but the ability to only show the label for the tile that is hovered over, similar to the `Picking` Sandcastle. A checkbox in the tile info tab can control which mode is being used.

**Comment by rahwang** [comment]
Hey @lilleyse , I redid my changes to reflect Hannah's recent refactoring of the inspector and addressed all the things you mentioned above (except the picking, which I'm working on). How should the tile info stats changes I've made so far be tested? Just follow the template of the debugShowGeometricError tests in the Cesium3DTilesetSpec?

**Comment by lilleyse** [comment]
Those tests can be renamed to `show tile info labels with _____`. In at least one of them all the debug options should be set to true.

Then a separate test can check how the labels change when new debug settings are toggled. Setting `viewRootOnly` at the beginning will keep the code more concise for that test.

**Comment by rahwang** [comment]
Ah, as per discussion with @lilleyse , going to refactor the debug label assembly to happen in the inspector view model instead of in the tileset.

**Comment by rahwang** [comment]
Ah, sorry about that @lilleyse -- I still had the original labels drawing for debugging purposes. Here is my properly broken code. XD

**Comment by lilleyse** [comment]
The problem is that the update is pushing commands after the render loop, and so they just get ignored. The same would happen will `preRender`. Maybe this means storing the label collection in the view model is  bad design.

**Comment by rahwang** [comment]
Ah, thanks for explaining, @lilleyse . So shall I just revert to the previous version?

**Comment by lilleyse** [comment]
Yeah, that's best for now.

**Comment by rahwang** [comment]
Ready for another look @lilleyse

**Comment by lilleyse** [comment]
Just to summarize the PR breakdown...

This one should be ready soon. Then open a 2nd PR to fix picking for tilesets that don't use a batch table. And then a 3rd PR to add this feature:

> the ability to only show the label for the tile that is hovered over, similar to the Picking Sandcastle. A checkbox in the tile info tab can control which mode is being used.

File: Source/Scene/Cesium3DTile.js [hunk 1]
```diff
@@ -851,9 +851,11 @@ define([
      * @private
      */
     Cesium3DTile.prototype.update = function(tileset, frameState) {
+        var initCommandLength = frameState.commandList.length;
         applyDebugSettings(this, tileset, frameState);
         this._content.update(tileset, frameState);
         this._transformDirty = false;
+        this._content.commandsLength = frameState.commandList.length - initCommandLength;
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTile.js]: It will be simpler if `Cesium3DTile` holds onto `this._commandsLength` instead of the contents.

File: Source/Scene/Cesium3DTileset.js [hunk 2]
```diff
@@ -401,7 +401,63 @@ define([
...
+         *
+         * @type {Boolean}
+         * @default false
+         */
+        this.debugShowNumberOfCommands = defaultValue(options.debugShowNumberOfCommands, false);
+
+        /**
+         * This property is for debugging only; it is not optimized for production use.
+         * <p>
+         * When true, draws labels to indicate the number of triangles in each tile.
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: triangles -> points

File: Source/Scene/Cesium3DTileset.js [hunk 3]
```diff
@@ -401,7 +401,63 @@ define([
...
+         * </p>
+         *
+         * @type {Boolean}
+         * @default false
+         */
+        this.debugShowNumberOfPoints = defaultValue(options.debugShowNumberOfPoints, false);
+
+
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Remove empty line.

File: Source/Scene/Cesium3DTileset.js [hunk 4]
```diff
@@ -1858,12 +1914,44 @@ define([
                 normal = Cartesian3.multiplyByScalar(normal, 0.75 * radius, scratchCartesian2);
                 position = Cartesian3.add(normal, boundingVolume.center, scratchCartesian2);
             }
-            tileset._geometricErrorLabels.add({
-                text: tile.geometricError.toString(),
-                position: position
+
+            var labelString = "";
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Style: Use single quotes instead double quotes throughout.

File: Source/Scene/Cesium3DTileset.js [hunk 5]
```diff
@@ -1858,12 +1914,44 @@ define([
...
+            var attribCounter = 0;
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Style: avoid abbreviations.

File: Source/Scene/Cesium3DTileset.js [hunk 6]
```diff
@@ -1858,12 +1914,44 @@ define([
...
+
+            if (tileset.debugShowGeometricError) {
+                labelString += "\nGeometric error: " + tile.geometricError.toString();
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: `toString()` for these isn't needed.

File: Source/Scene/Cesium3DTileset.js [hunk 7]
```diff
@@ -1858,12 +1914,44 @@ define([
...
+            }
+            if (tileset.debugShowVertexMemoryUsage) {
+                labelString += "\nVertex Memory: " + tile.content.vertexMemorySizeInBytes.toString();
+                attribCounter++;
+            }
+
+            tileset._tileInfoLabels.add({
+                text: labelString.substring(1),
+                position: position,
+                font : (18-attribCounter).toString() + 'px sans-serif',
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: `toString()` not needed

File: Source/Scene/Cesium3DTileset.js [hunk 8]
```diff
@@ -1858,12 +1914,44 @@ define([
...
+                labelString += "\nTexture Memory: " + tile.content.textureMemorySizeInBytes.toString();
+                attribCounter++;
+            }
+            if (tileset.debugShowVertexMemoryUsage) {
+                labelString += "\nVertex Memory: " + tile.content.vertexMemorySizeInBytes.toString();
+                attribCounter++;
+            }
+
+            tileset._tileInfoLabels.add({
+                text: labelString.substring(1),
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Be consistent with the :'s. Usually the convention is space on both sides (`text : labelString.substring(1),`)

File: Source/Scene/Cesium3DTileset.js [hunk 9]
```diff
@@ -1948,13 +2036,14 @@ define([
...
+        if (tileset.debugShowGeometricError || tileset.debugShowNumberOfCommands || tileset.debugShowTextureMemoryUsage ||
+            tileset.debugShowVertexMemoryUsage || tileset.debugShowNumberOfPoints || tileset.debugShowNumberOfTriangles) {
+            if (!defined(tileset._tileInfoLabels)) {
+                tileset._tileInfoLabels = new LabelCollection();
             }
-            updateGeometricErrorLabels(tileset, frameState);
+            updateTileInfoLabels(tileset, frameState);
         } else {
-            tileset._geometricErrorLabels = tileset._geometricErrorLabels && tileset._geometricErrorLabels.destroy();
+            tileset._tileInfoLabels = tileset._tileInfoLabels && tileset._tileInfoLabels.destroy();
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Also destroy the labels in `destroy`.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspector.js [hunk 10]
```diff
@@ -196,18 +196,27 @@ define([
         errorBox.setAttribute('data-bind', 'text: editorError');
         stylePanelContents.appendChild(errorBox);
 
+        tileInfoPanelContents.appendChild(makeCheckbox('showGeometricError', 'Geometric Error'));
+        tileInfoPanelContents.appendChild(makeCheckbox('textureMemory', 'Texture Memory Usage'));
+        tileInfoPanelContents.appendChild(makeCheckbox('vertexMemory', 'Vertex Memory Usage'));
+        tileInfoPanelContents.appendChild(makeCheckbox('numberOfCommands', 'Number of Commands'));
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspector.js]: Captialize `of` just to be consistent with the two below.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspector.js [hunk 11]
```diff
@@ -196,18 +196,27 @@ define([
...
         var stylePanel = makeSection('Style', 'styleVisible', 'toggleStyle', stylePanelContents);
+        var tileInfoPanel = makeSection('Tile Info', 'tileInfoVisible', 'toggleTileInfo', tileInfoPanelContents);
 
         // first add and bind all the toggleable panels
         element.appendChild(tilesetPanel);
         element.appendChild(displayPanel);
         element.appendChild(updatePanel);
         element.appendChild(loggingPanel);
         element.appendChild(stylePanel);
+        element.appendChild(tileInfoPanel);
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspector.js]: Visually, I think the tile info panel belongs after the logging panel.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js [hunk 12]
```diff
@@ -317,6 +326,8 @@ define([
                 }
             }
         });
+
+
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: Remove empty lines.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js [hunk 13]
```diff
@@ -469,6 +480,132 @@ define([
...
+            get : function() {
+                return onlyPickedTileInfo();
+            },
+            set : function(val) {
+                onlyPickedTileInfo(val);
+                if (that._tileset) {
+                    that._tileset.debugShowOnlyPickedTile = val;
+                }
+            }
+        });
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: As mentioned in my other comment, the per-tile picking could be saved for another PR.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js [hunk 14]
```diff
@@ -223,6 +223,15 @@ define([
...
 
         /**
+         * Gets or sets the flag to show the tile info section.  This property is observable.
+         *
+         * @type {Boolean}
+         * @default false
+         */
+        this.tileInfoVisible = false;
+
+
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: Remove empty line.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js [hunk 15]
```diff
@@ -469,6 +480,132 @@ define([
...
+            }
+        });
+        /**
+         * Displays the texture memory used per tile.  This property is observable.
+         * @memberof Cesium3DTilesInspectorViewModel.prototype
+         *
+         * @type {Boolean}
+         * @default false
+         */
+        this.textureMemory = false;
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: Place the `showGeometricError` stuff closer to these.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js [hunk 16]
```diff
@@ -469,6 +480,132 @@ define([
...
+
+        var numberOfTriangles = knockout.observable();
+        knockout.defineProperty(this, 'numberOfTriangles', {
+            get : function() {
+                return numberOfTriangles();
+            },
+            set : function(val) {
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: Change all `val` to `value`.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js (hunk 15 above)
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: Also it would be better to arrange these in the order they are shown in the UI.

File: Specs/Scene/Cesium3DTilesetSpec.js [hunk 17]
```diff
@@ -1315,58 +1315,201 @@ defineSuite([
...
-            expect(tileset._geometricErrorLabels._labels[0].text).toEqual('70');
-            expect(tileset._geometricErrorLabels._labels[1].text).toEqual('0');
-            expect(tileset._geometricErrorLabels._labels[2].text).toEqual('0');
-            expect(tileset._geometricErrorLabels._labels[3].text).toEqual('0');
-            expect(tileset._geometricErrorLabels._labels[4].text).toEqual('0');
+            expect(tileset._tileInfoLabels).toBeDefined();
+            expect(tileset._tileInfoLabels.length).toEqual(5);
+
+            var root = tileset._root;
+            expect(tileset._tileInfoLabels._labels[0].text).toEqual('Geometric error: ' + root.geometricError.toString());
```
Comment (lilleyse) [inline Specs/Scene/Cesium3DTilesetSpec.js]: Remove uses of `toString()` throughout this file.

File: Specs/Scene/Cesium3DTilesetSpec.js [hunk 18]
```diff
@@ -1315,58 +1315,201 @@ defineSuite([
...
+            scene.renderForSpecs();
+            expect(tileset._tileInfoLabels).toBeDefined();
+
+            var content = tileset._root.content;
+            expect(tileset._tileInfoLabels._labels[0].text).toEqual('Geometric error: ' + tileset._root.geometricError.toString() + '\n' +
+                                                                    'Commands: ' + content.commandsLength.toString() + '\n' +
+                                                                    'Points: ' + content.pointsLength.toString() + '\n' +
+                                                                    'Triangles: ' + content.trianglesLength.toString() + '\n' +
+                                                                    'Texture Memory: ' + content.textureMemorySizeInBytes.toString() + '\n' +
+                                                                    'Vertex Memory: ' + content.vertexMemorySizeInBytes.toString());
```
Comment (lilleyse) [inline Specs/Scene/Cesium3DTilesetSpec.js]: To avoid the large indentation, create a var for the string.

**Comment by lilleyse** [comment]
Fixes https://github.com/AnalyticalGraphicsInc/cesium/issues/5029 too
I forgot that issue existed...

@rahwang can you also add `Number of features` to this PR?

**Comment by rahwang** [comment]
Sure! I'll do that now.

**Comment by rahwang** [comment]
Ok @lilleyse, I added feature counts and updated according to your review. Ready for another look.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspector.js [hunk 19]
```diff
@@ -132,6 +132,7 @@ define([
         var updatePanelContents = document.createElement('div');
         var loggingPanelContents = document.createElement('div');
         var stylePanelContents = document.createElement('div');
+        var tileInfoPanelContents = document.createElement('div');
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspector.js]: Also place this before stylePanel for consistency.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js [hunk 20]
```diff
@@ -448,26 +478,131 @@ define([
...
+         * Displays the texture memory used per tile.  This property is observable.
+         * @memberof Cesium3DTilesInspectorViewModel.prototype
          *
          * @type {Boolean}
          * @default false
          */
-        this.freezeFrame = false;
+        this.textureMemory = false;
+
+        var vertexMemory = knockout.observable();
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: I missed this from before. Can you rename all the new properties to `showTextureMemory` to follow the `showGeometricError` convention?

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js [hunk 21]
```diff
@@ -681,7 +817,13 @@ define([
                                     'showContentBoundingVolumes',
                                     'showRequestVolumes',
                                     'showGeometricError',
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: Move `showGeometricError` below near the others.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js [hunk 22]
```diff
@@ -681,7 +817,13 @@ define([
...
-                                    'freezeFrame'];
+                                    'freezeFrame',
+                                    'onlyPickedTileInfo',
+                                    'textureMemory',
+                                    'vertexMemory',
+                                    'numberOfPoints',
+                                    'numberOfTriangles',
+                                    'numberOfCommands'];
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: Missing `numberOfFeatures`.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js [hunk 23]
```diff
@@ -681,7 +817,13 @@ define([
                                     'showContentBoundingVolumes',
                                     'showRequestVolumes',
                                     'showGeometricError',
-                                    'freezeFrame'];
+                                    'freezeFrame',
+                                    'onlyPickedTileInfo',
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: Remove

File: Source/Scene/Cesium3DTileset.js [hunk 24]
```diff
@@ -1858,12 +1924,48 @@ define([
...
+            if (tileset.debugShowNumberOfFeatures) {
+                labelString += "\nFeatures: " + tile.content.featuresLength;
+                attributes++;
+            }
+            if (tileset.debugShowTextureMemoryUsage) {
+                labelString += "\nTexture Memory: " + tile.content.textureMemorySizeInBytes;
+                attributes++;
+            }
+            if (tileset.debugShowVertexMemoryUsage) {
+                labelString += "\nVertex Memory: " + tile.content.vertexMemorySizeInBytes;
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Late suggestion, but can this and texture memory usage show it in megabytes like in the stats panel in the inspector?

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js (hunk 20 above)
Comment (rahwang) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: You didn't miss it! I accidentally reverted it along with the labels-to-view-model commits.

**Comment by rahwang** [comment]
Thanks for the review @lilleyse ! Comments addressed.

File: Source/Scene/Cesium3DTileset.js [hunk 25]
```diff
@@ -2221,7 +2221,7 @@ define([
     Cesium3DTileset.prototype.destroy = function() {
         // Destroy debug labels
         if (defined(this._tileInfoLabels)) {
-            this._tileInfoLabels.destroy();
+            this._tileInfoLabels = this._tileInfoLabels && this._tileInfoLabels.destroy();
         }
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Remove the `if (defined(this._tileInfoLabels)) {` wrapper, it is unnecessary now.

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspector.js [hunk 26]
```diff
@@ -196,17 +196,27 @@ define([
         errorBox.setAttribute('data-bind', 'text: editorError');
         stylePanelContents.appendChild(errorBox);
 
+        tileInfoPanelContents.appendChild(makeCheckbox('showGeometricError', 'Geometric Error'));
+        tileInfoPanelContents.appendChild(makeCheckbox('showNumberOfCommands', 'Number of Commands'));
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspector.js]: Change to `Number Of Commands`

File: Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js [hunk 27]
```diff
@@ -573,8 +708,9 @@ define([
         this._style = undefined;
         this._shouldStyle = false;
         this._definedProperties = ['properties', 'dynamicScreenSpaceError', 'colorBlendMode', 'picking', 'colorize', 'wireframe', 'showBoundingVolumes',
-                                   'showContentBoundingVolumes', 'showRequestVolumes', 'showGeometricError', 'freezeFrame', 'maximumScreenSpaceError',
-                                   'dynamicScreenSpaceErrorDensity', 'dynamicScreenSpaceErrorDensitySliderValue', 'dynamicScreenSpaceErrorFactor', 'pickActive'];
+                                   'showContentBoundingVolumes', 'showRequestVolumes', 'freezeFrame', 'maximumScreenSpaceError', 'dynamicScreenSpaceErrorDensity',
+                                   'dynamicScreenSpaceErrorDensitySliderValue', 'dynamicScreenSpaceErrorFactor', 'pickActive', 'showGeometricError',
+                                   'showTextureMemory', 'showVertexMemory', 'showNumberOfCommands', 'showNumberOfPoints', 'showNumberOfTriangles', 'showNumberOfFeatures'];
```
Comment (lilleyse) [inline Source/Widgets/Cesium3DTilesInspector/Cesium3DTilesInspectorViewModel.js]: Also reorder them here to match.

File: Source/Scene/Cesium3DTileset.js [hunk 28]
```diff
@@ -401,7 +401,73 @@ define([
...
+        /**
+         * This property is for debugging only; it is not optimized for production use.
+         * <p>
+         * When true, draws labels to indicate the vertex memory usage.
+         * </p>
+         *
+         * @type {Boolean}
+         * @default false
+         */
+        this.debugShowVertexMemoryUsage = defaultValue(options.debugShowVertexMemoryUsage, false);
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Can you add these `debugShow` options to the doc in the constructor?

**Comment by lilleyse** [comment]
Ok I think these are the last points, almost ready!

**Comment by rahwang** [comment]
Done, @lilleyse!

File: Source/Scene/Cesium3DTileset.js [hunk 29]
```diff
@@ -121,7 +121,13 @@ define([
      * @param {Boolean} [options.debugShowBoundingVolume=false] For debugging only. When true, renders the bounding volume for each tile.
      * @param {Boolean} [options.debugShowContentBoundingVolume=false] For debugging only. When true, renders the bounding volume for each tile's content.
      * @param {Boolean} [options.debugShowViewerRequestVolume=false] For debugging only. When true, renders the viewer request volume for each tile.
-     * @param {Boolean} [options.debugShowGeometricError=false] For debugging only. When true, draws labels to indicate the geometric error of each tile
+     * @param {Boolean} [options.debugShowGeometricError=false] For debugging only. When true, draws labels to indicate the geometric error of each tile.
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: These are probably too fine-grained.  Consider:

* Combine `debugShowNumberOfCommands`,  `debugShowNumberOfPoints`, `debugShowNumberOfTriangles`, and `debugShowNumberOfFeatures` into `debugShowRenderingStatistics`
* Combine `debugShowTextureMemoryUsage` and `debugShowVertexMemoryUsage` into `debugShowMemoryUsage`

This will be a lot less code and is inline with how I expect folks will want to use these hopefully without cluttering the display.

If clutter is a problem, try a smaller font that is still readable or a slightly less coarse-grained separation for the rendering stats.

File: Source/Scene/Cesium3DTileset.js [hunk 30]
```diff
@@ -401,7 +407,73 @@ define([
          * @default false
          */
         this.debugShowGeometricError = defaultValue(options.debugShowGeometricError, false);
-        this._geometricErrorLabels = undefined;
+        this._tileInfoLabels = undefined;
+
+        /**
+         * This property is for debugging only; it is not optimized for production use.
+         * <p>
+         * When true, draws labels to indicate the number of commands used.
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: "commands used" -> "commands executed for this tile"

Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: "each tile"

**Comment by lilleyse** [comment]
If going with the combined approach, it would make sense to hide `number of triangles` and `number of points` when those values are 0.

**Comment by rahwang** [comment]
Thanks for the suggestions, @pjcozzi and @lilleyse. Fixed!

File: Source/Scene/Cesium3DTileset.js [hunk 31]
```diff
@@ -1858,12 +1882,49 @@ define([
...
+                    attributes++;
+                }
+                var numberOfTriangles = tile.content.trianglesLength;
+                if (numberOfTriangles > 0) {
+                    labelString += '\nTriangles: ' + tile.content.trianglesLength;
+                    attributes++;
+                }
+
+                labelString += '\nFeatures: ' + tile.content.featuresLength;
+                attributes += 4;
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: This should be ++ now.

Comment (rahwang) [inline Source/Scene/Cesium3DTileset.js]: whoops! Fixed @lilleyse

**Comment by lilleyse** [comment]
Nice work @rahwang.

File: Source/Scene/Cesium3DTile.js [hunk 32]
```diff
@@ -142,6 +142,8 @@ define([
         }
         this._viewerRequestVolume = viewerRequestVolume;
 
+        this._commandsLength = 0;
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTile.js]: This should have a readonly property getter.

Up to you, but this and the other "length" properties for rendering debugging output could be prefixed like `_debugCommandsLength`

File: Source/Scene/Cesium3DTileset.js [hunk 33]
```diff
@@ -1838,10 +1862,10 @@ define([
 
     var scratchCartesian2 = new Cartesian3();
 
-    function updateGeometricErrorLabels(tileset, frameState) {
+    function updateTileInfoLabels(tileset, frameState) {
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: `TileInfo` is very generic; we try to avoid names like "info."  Perhaps `updateTileDebugLabels`.

Same for the `_tileInfoLabels` naming.

Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: `tileInfoVisible`, etc.
